Sso node walkthrough (#1442)
OfficeDev/office-js-docs-pr: docs/develop/create-sso-office-add-ins-nodejs.md

1. Open the add-in manifest file “manifest\manifest_local.xml” and then scroll to the bottom of the file. Just above the end `</VersionOverrides>` tag, you'll find the following markup:

    ```xml
    <WebApplicationInfo>
      <Id>$application_GUID here$</Id>
      <Resource>api://localhost:44355/$application_GUID here$</Resource>
      <Scopes>
          <Scope>Files.Read.All</Scope>
          <Scope>profile</Scope>
      </Scopes>
    </WebApplicationInfo>
    ```

1. Replace the placeholder “$application_GUID here$” *in both places* in the markup with the Application ID that you copied when you registered your add-in. The "$" symbols are not part of the ID, so do not include them. This is the same ID you used in for the ClientID and Audience in the web.config.

	> [!NOTE]
    > The **Resource** value is the **Application ID URI** you set when you registered the add-in. The **Scopes** section is used only to generate a consent dialog box if the add-in is sold through AppSource.

## Code the client-side

### Create the SSO logic

> 1. In your code editor, open the file `public\javascripts\ssoAuthES6.js`. It already has code that ensures that Promises are supported, even in Internet Explorer 11, and a call of Office.onReady to assign a handler to the add-in's only button.

Reviewer: ```suggestion
1. In your code editor, open the file `public\javascripts\ssoAuthES6.js`. It already has code that ensures that Promises are supported, even in Internet Explorer 11, and an `Office.onReady` call to assign a handler to the add-in's only button.
```

	> [!NOTE]
    > As the name suggests, the ssoAuthES6.js uses JavaScript ES6 syntax because using `async` and `await` best shows the essential simplicity of the SSO API. When the localhost server is started, this file is transpiled to ES5 syntax so that the sample will run in Internet Explorer 11. 

1. Add the following code below the Office.onReady method:

    ```javascript
    async function getGraphData() {
        try {
            
            // TODO 1: Tell Office to get a bootstrap token from Azure AD.
            
            // TODO 2: Attempt to exhange the bootstrap token for an 
            //         access token to Microsoft Graph.

            // TODO 3: Handle case where Microsoft Graph requires an 
            //         additional form of authentication.

            // TODO 4: Use the access token in a call to Microsoft Graph 
            //         or handle any error from the attempted token exchange.

        }
        catch(exception) {

>             // TODO 5: Respond to exceptions thrown by the call of 

Reviewer: ```suggestion
            // TODO 5: Respond to exceptions thrown by the
```

>             //         OfficeRuntime.auth.getAccessToken.

Reviewer: ```suggestion
            //         OfficeRuntime.auth.getAccessToken call.
```

        }
    }
    ```

1. Replace `TODO 1` with the following code. About this code, note:

    - `OfficeRuntime.auth.getAccessToken` instructs Office to get a bootstrap token from Azure AD. A bootstrap token is similar to an ID token, but it has a `scp` (scope) property with the value `access-as-user`. This kind of token can be exchanged by a web application for an access token to Microsoft Graph.

>     - Setting the `allowSignInPrompt`option to true means that if no user is currently signed in to Office, then Office will open a popup signin prompt.

Reviewer: ```suggestion
    - Setting the `allowSignInPrompt`option to true means that if no user is currently signed into Office, then Office will open a popup sign-in prompt.
```

>     - Setting the `forMSGraphAccess` option to true signals to Office that the add-in intends to use the bootstrap token to get an access token to Micrsoft Graph, instead of just using it as an ID token. If the tenant administrator has not granted consent to the add-in's access to Microsoft Graph, then `OfficeRuntime.auth.getAccessToken` returns error **13012**. The add-in can respond by falling back to an alternative system of authorization, which is necessary because Office can prompt only for consent to the user's Azure AD profile, not to any Microsoft Graph scopes. The fallback authorization system requires the user to sign-in again and the user *can* be prompted to consent to Micrsoft Graph scopes. So, the `forMSGraphAccess` option ensures that the add-in won't make a token exchange that will fail due to lack of consent. (Since you granted administrator consent in an earlier step, this scenario won't happen for this add-in. But the option is included here anyway to illustrate a best practice.)

Reviewer: ```suggestion
    - Setting the `forMSGraphAccess` option to true signals to Office that the add-in intends to use the bootstrap token to get an access token to Micrsoft Graph, instead of just using it as an ID token. If the tenant administrator has not granted consent to the add-in's access to Microsoft Graph, then `OfficeRuntime.auth.getAccessToken` returns error **13012**. The add-in can respond by falling back to an alternative system of authorization, which is necessary because Office can prompt only for consent to the user's Azure AD profile, not to any Microsoft Graph scopes. The fallback authorization system requires the user to sign in again and the user *can* be prompted to consent to Micrsoft Graph scopes. So, the `forMSGraphAccess` option ensures that the add-in won't make a token exchange that will fail due to lack of consent. (Since you granted administrator consent in an earlier step, this scenario won't happen for this add-in. But the option is included here anyway to illustrate a best practice.)
```

    ```javascript
    let bootstrapToken = await OfficeRuntime.auth.getAccessToken({ allowSignInPrompt: true, forMSGraphAccess: true }); 
    ```

1. Replace `TODO 2` with the following code. You'll create the `getGraphToken` method in a later step.

    ```javascript
    let exchangeResponse = await getGraphToken(bootstrapToken);
    ```

1. Replace `TODO 3` with the following. About this code, note: 

>     - If the Office 365 tenant has been configured to require multifactor authentication, then the `exchangeResponse` will include a `claims` property with information about the additional required factors. In that case, `OfficeRuntime.auth.getAccessToken` should be recalled with the `authChallenge` option set to the value of the claims property. This tells AAD to prompt the user for all required forms of authentication.

Reviewer: ```suggestion
    - If the Office 365 tenant has been configured to require multifactor authentication, then the `exchangeResponse` will include a `claims` property with information about the additional required factors. In that case, `OfficeRuntime.auth.getAccessToken` should be called again with the `authChallenge` option set to the value of the claims property. This tells AAD to prompt the user for all required forms of authentication.
```

    ```javascript
    if (exchangeResponse.claims) {
        let mfaBootstrapToken = await OfficeRuntime.auth.getAccessToken({ authChallenge: exchangeResponse.claims });
        exchangeResponse = await getGraphToken(mfaBootstrapToken);
    }
    ```

> 1. Replace `TODO 4` with the following. About this code,note: 

Reviewer: ```suggestion
1. Replace `TODO 4` with the following. About this code, note: 
```

>     - You'll create the `handleAADErrors` method in a later step. Azure AD errors are returned to the client as HTTP code 200 Responses. They do not throw errors, so they do not trigger the catch block of the `getGraphData` method.

Reviewer: ```suggestion
    - You'll create the `handleAADErrors` method in a later step. Azure AD errors are returned to the client as HTTP code 200 Responses. They do not throw errors, so they do not trigger the `catch` block of the `getGraphData` method.
```

>     - You'll create the `makeGraphApiCall` method in a later step. It makes an AJAX call to the MS Graph endpoint. Errors are caught in the `.fail` callback of that call, not in the catch block of the `getGraphData` method.

Reviewer: ```suggestion
    - You'll create the `makeGraphApiCall` method in a later step. It makes an AJAX call to the MS Graph endpoint. Errors are caught in the `.fail` callback of that call, not in the `catch` block of the `getGraphData` method.
```

        // No one is signed into Office. If the add-in cannot be effectively used when no one 
        // is logged into Office, then the first call of getAccessToken should pass the 
        // `allowSignInPrompt: true` option. Since this add-in does that, you should not see
        // this error. 
        showMessage("No one is signed into Office. But you can use many of the add-ins functions anyway. If you want to log in, press the Get OneDrive File Names button again.");  
        break;
    case 13002:
        // OfficeRuntime.auth.getAccessToken was called with the allowConsentPrompt 
        // option set to true. But, the user aborted the consent prompt. 
        showMessage("You can use many of the add-ins functions even though you have not granted consent. If you want to grant consent, press the Get OneDrive File Names button again."); 
        break;
    case 13006:
        // Only seen in Office on the Web.
        showMessage("Office on the Web is experiencing a problem. Please sign out of Office, close the browser, and then start again."); 
        break;
    case 13008:
        // The OfficeRuntime.auth.getAccessToken method has already been called and 

>         // that call has not completed yet. Only seen in Office on the Web.

Reviewer: ```suggestion
        // that call has not completed yet. Only seen in Office on the web.
```

        showMessage("Office is still working on the last operation. When it completes, try this operation again."); 
        break;
    case 13010:

>         // Only seen in Office on the Web.

Reviewer: ```suggestion
        // Only seen in Office on the web.
```

        showMessage("Follow the instructions to change your browser's zone configuration.");
        break;
    ```

> 1. Replace `TODO 7` with the following code. or more information about these errors, see [Troubleshoot SSO in Office Add-ins](troubleshoot-sso-in-office-add-ins.md). The function `dialogFallback` invokes the alternative system of authorization. In this add-in, the fallback system opens a dialog which requires the user to sign in, even if the user already is, and uses msal.js and the Implicit Flow to get an access token to Microsoft Graph.

Reviewer: ```suggestion
1. Replace `TODO 7` with the following code. For more information about these errors, see [Troubleshoot SSO in Office Add-ins](troubleshoot-sso-in-office-add-ins.md). The function `dialogFallback` invokes the alternative system of authorization. In this add-in, the fallback system opens a dialog which requires the user to sign in, even if the user already is, and uses msal.js and the Implicit Flow to get an access token to Microsoft Graph.
```

    ```javascript
    default:
    // For all other errors, including 13000, 13003, 13005, 13007, 13012, 
    // and 50001, fall back to non-SSO sign-in.
    dialogFallback();
    break;
    ```

1. Below the `handleClientSideErrors` function, add the following function. 

    ```javascript

>     function handleAADErrors(exchangeResponse) {

Reviewer: ```suggestion
    function handleAzureADErrors(exchangeResponse) {
```

Author: Changing the code would be difficult and I don't think there's a problem with abbreviating "Azure" in a function name.

    // TODO 8: Handle case where the bootstrap token is expired.

>     // TODO 9: Handle all other AAD errors.

Reviewer: ```suggestion
    // TODO 9: Handle all other Azure AD errors.
```

    
    }
    ```

> 1. On rare occasions the bootstrap token that Office has cached is unexpired when Office validates it, but expires by the time it reaches Azure AD for exchange. Azure AD will respond with error **AADSTS500133**. In this case, the add-in should simply recursively recall `getGraphData`. Since the cached bootstrap token is now expired, Office will get a new one from Azure AD. The So replace `TODO 8` with the following. 

Reviewer: ```suggestion
1. On rare occasions the bootstrap token that Office has cached is unexpired when Office validates it, but expires by the time it reaches Azure AD for exchange. Azure AD will respond with error **AADSTS500133**. In this case, the add-in should simply recursively call `getGraphData`. Since the cached bootstrap token is now expired, Office will get a new one from Azure AD. So replace `TODO 8` with the following. 
```

    ```javascript
    if (exchangeResponse.error_description.indexOf("AADSTS500133") !== -1)       
    {
        getGraphData();
    }
    ```

> 1. To ensure that the add-in doesn't enter an infinite loop of calls to `getGraphData`, the add-in should keep track of how many times `getGraphData` has been called and be sure that is not called recursively called more than once. So, create a counter variable in a scope that is global to the `handleAADErrors` and `getGraphData` functions. A good place for global variables is just below the call of the `Office.onReady` method. 

Reviewer: ```suggestion
1. To ensure that the add-in doesn't enter an infinite loop of calls to `getGraphData`, the add-in should keep track of how many times `getGraphData` has been called and be sure that is not called recursively called more than once. So, create a counter variable in a scope that is global to the `handleAADErrors` and `getGraphData` functions. A good place for global variables is just below the `Office.onReady` method call.
```

        &&
        (retryGetAccessToken <= 0)) 
    {
        retryGetAccessToken++;
        getGraphData();
    }
    ```

1. Replace `TODO 9` with the following. 

    ```javascript
    else 
    {                
        dialogFallback();
    }
    ```

1. Save and close the file.

### Get the data and add it to the Office document

1. In the `public\javascripts` folder, create a new file named `data.js`.

1. Add the following function to the file. This is the function that is called by the `getGraphData` function when it has acquired an access token to Microsoft Graph. 

    ```javascript
    function makeGraphApiCall(accessToken) {
        $.ajax(

            // TODO 10: Call an Express route on the add-in's server-side 
            //          code and pass the access token to Microsoft Graph.

        )
        .done(function (response) {

            // TODO 11: Write the data received from Microsoft Graph to 
            //          the Office document.

        })
        .fail(function (errorResult) {
            showMessage("Error from Microsoft Graph: " + JSON.stringify(errorResult));
        });
    }
    ```

> 1. Replace `TODO 10` woth the following. About this code, note: 

Reviewer: ```suggestion
1. Replace `TODO 10` with the following. About this code, note: 
```

    - This object is the parameter to the `$.ajax` method.

>     - The `/getuserdata` is an Express route on the add-in's server that you create in a later step. It will call a Microsoft Graph endpoint and include the access token in it's call. 

Reviewer: ```suggestion
    - The `/getuserdata` is an Express route on the add-in's server that you create in a later step. It will call a Microsoft Graph endpoint and include the access token in its call. 
```

    ```javascript
    {
        type: "GET", 
        url: "/getuserdata",
        headers: {"access_token": accessToken },
        cache: false
    }
    ```

1. Replace `TODO11` with the following. About this code, note:

>     - The `writeFileNamesToOfficeDocument` will insert the data from graph into the Office document. It is defined in the `public\javascripts\document.js` file. 

Reviewer: ```suggestion
    - The `writeFileNamesToOfficeDocument` will insert the data from Graph into the Office document. It is defined in the `public\javascripts\document.js` file. 
```

>     - If `writeFileNamesToOfficeDocument` returns an error, it will begin "Unable to add filenames to document."

Reviewer: ```suggestion
    - If `writeFileNamesToOfficeDocument` returns an error, it will begin with "Unable to add filenames to document."
```

        // TODO 12: Test for the presence of the Authorization header.

        // TODO 13: Create the hidden form that will be sent to Azure AD 
        //          to request the access token in exhange for the 
        //          bootstrap token.

        // TODO 14: Send the POST request to Azure AD and relay the 
        //          access token (or an error) to the client.

    });
    ```

1. Replace `TODO 12` with the following code.

    ```javascript
    const authorization = req.get('Authorization');
    if (authorization == null) {
        let error = new Error('No Authorization header was found.');
        next(error);
    } 
    ```

1. Replace `TODO 13` with the following code. About this code, note: 

    - This is the beginning of a long `else` block, but the closing `}` is not at the end yet because you will be adding more code to it. 
    - The `authorization` string is "Bearer " followed by the bootstrap token, so the first line of the `else` block is assigning the token to the `jwt`. ("JWT" stands for "JSON Web Token".)
    - The two `process.env.*` values are the constants that you assigned when you configured the add-in. 

>     - The `requested_token_use` form parameter is set to 'on_behalf_of'. This tells Azure AD that the add-in is requesting an access token to Microsoft Graph using the On-Behalf-Of Flow. Azure will respond by validating that the bootstrap token, which is assigned to`assertion` form parameter, has a `scp` property that is set to `access-as-user`.

Reviewer: ```suggestion
    - The `requested_token_use` form parameter is set to 'on_behalf_of'. This tells Azure AD that the add-in is requesting an access token to Microsoft Graph using the On-Behalf-Of Flow. Azure will respond by validating that the bootstrap token, which is assigned to `assertion` form parameter, has a `scp` property that is set to `access-as-user`.
```

        //          to the client.
        
    });
    ```

1. Replace `TODO 15` with the following. About this code, note:

    - The caller of this route, `makeGraphApiCall`, added the access token to Microsoft Graph to the HTTP Request as a header named "access_token".
    - The `getGraphData` function is defined in the `msgraph-helper.js` file. (This is not the same function as the client-side `getGraphData` function that you defined in the `ssoAuthES6.js` file.)
    - The last parameter, for `queryParamsSegment`, is hardcoded. If you reuse this code in a production add-in and any part of `queryParamsSegment` comes from user input, be sure that it is sanitized so that it cannot be used in a Response header injection attack.
    - The code minimizes the data that must come from Microsoft Graph by specifying only the property we need ("name") and only the top 10 folder or file names.

    ```javascript
    const graphToken = req.get('access_token');    
    const graphData = await getGraphData(graphToken, "/me/drive/root/children", "?$select=name&$top=10");
    ```

1. Replace `TODO 16` with the following. About this code, note:

>     - If Microsoft Graph returns an error, such as invalid or expired token, there will be a code property in the returned object set to a HTTP status (e.g. 401). The code relays the error to the client. It will br caught in the `.fail` callback of `makeGraphApiCall`.

Reviewer: ```suggestion
    - If Microsoft Graph returns an error, such as invalid or expired token, there will be a code property in the returned object set to a HTTP status (e.g., 401). The code relays the error to the client. It will be caught in the `.fail` callback of `makeGraphApiCall`.
```

    - Microsoft Graph data includes OData metadata and eTags that the add-in does not need, so the code constructs a new array containing only the file names to send to the client.

    ```javascript
    if (graphData.code) {

>         next(createError(graphData.code, "Microsoft Graph error " + JSON.stringify(graphData)));

Reviewer: ```suggestion
        next(createError(graphData.code, "Microsoft Graph error: " + JSON.stringify(graphData)));
```

    }

>     else 

Reviewer: I noticed in some cases you have `else {` but in other cases you split it into separate lines. I recommend they should be consistent. My preference is `else {`.

    {
        const itemNames = [];
        const oneDriveItems = graphData['value'];
        for (let item of oneDriveItems) {
            itemNames.push(item['name']);
        }

        res.send(itemNames)
    }
    ```

1. Save and close the file.

## Run the project

1. Ensure that you have some files in your OneDrive so that you can verify the results.

> 1. Open a command prompt in the root of the root of the `\Complete` folder. 

Reviewer: ```suggestion
1. Open a command prompt in the root of the `\Complete` folder. 
```

1. Run the command `npm start`. 

> 1. You need to sideload the add-in into an Office application (Excel, Word, or PowerPoint) to test it. The instructions depend on your platform. There are links to instructions at [Sideload an Office Add-in for Testing](https://docs.microsoft.com/office/dev/add-ins/testing/test-debug-office-add-ins#sideload-an-office-add-in-for-testing).

Reviewer: ```suggestion
1. You need to sideload the add-in into an Office application (Excel, Word, or PowerPoint) to test it. The instructions depend on your platform. There are links to instructions at [Sideload an Office Add-in for Testing](../testing/test-debug-office-add-ins.md#sideload-an-office-add-in-for-testing).
```

1. In the Office application, on the **Home** ribbon, select the **Show Add-in** button in the **SSO Node.js** group to open the task pane add-in.

1. Click the **Get OneDrive File Names** button. If you are logged into Office with either a Work or School (Office 365) account or Microsoft Account, and SSO is working as expected, the first 10 file and folder names in your OneDrive for Business are inserted into the document. (It may take as much as 15 seconds the first time.) If you are not logged in, or you are in a scenario that does not support SSO, or SSO is not working for any reason, you will be prompted to log in. After you log in, the file and folder names appear.

> [!NOTE]

> > If you were previously signed on to Office with a different ID, and some Office applications that were open at the time are still open, Office may not reliably change your ID even if it appears to have done so. If this happens, the call to Microsoft Graph may fail or data from the previous ID may be returned. To prevent this, be sure to *close all other Office applications* before you press **Get OneDrive File Names**.

Reviewer: ```suggestion
> If you were previously signed into Office with a different ID, and some Office applications that were open at the time are still open, Office may not reliably change your ID even if it appears to have done so. If this happens, the call to Microsoft Graph may fail or data from the previous ID may be returned. To prevent this, be sure to *close all other Office applications* before you press **Get OneDrive File Names**.
```
